tests/integration/engines/test_engine_table.py

```diff
@@ -0,0 +1,414 @@
+    col = get_func("col", session)
+    expr = get_func("expr", session)
+    lit = get_func("lit", session)
+    if isinstance(
+        session,
+        (
+            StandaloneSession,
+            PySparkSession,
+            RedshiftSession,
+            SnowflakeSession,
+            SparkSession,
+            BigQuerySession,
+            DuckDBSession,
+        ),
+    ):
+        pytest.skip("Engine doesn't support merge")
+
+    if isinstance(session, DatabricksSession):
+        uuid_func = "uuid"
+    elif isinstance(session, PostgresSession):
+        uuid_func = "gen_random_uuid"
+    else:
+        pytest.skip("Cannot generate uuids in this engine")
+
+    data, schema = merge_data
+    df_employee = session.createDataFrame(data, schema)
+    df_employee.write.saveAsTable("merge_employee")
+    df = session.read.table("merge_employee")
+    assert isinstance(df, _BaseTable)
+
+    start_date = "2024-01-01"
+    end_date = "2900-01-01"
+    df2 = session.createDataFrame(
+        [
+            (1, "Jack", "Shephard", 38, 2),
+            (6, "Mary", "Sue", 21, 45),
+        ],
+        ["employee_id", "fname", "lname", "age", "store_id"],
+    )
+
+    source = df2.select(
+        expr(f"{uuid_func}()").alias("s_id"),
+        col("employee_id").alias("employee_id"),
+        col("fname").alias("fname"),
+        col("lname").alias("lname"),
+        col("age").alias("age"),
+        col("store_id").alias("store_id"),
+        lit(end_date).alias("end_date"),
+    )
+
+    updates = (
+        source.alias("src")
+        .join(
+            df.alias("tg"),
+            on=(col("tg.employee_id") == col("src.employee_id")),
+        )
+        .where(
+            f"tg.end_date = '{end_date}' "
+            f"AND ("
+            f"src.fname <> tg.fname OR src.lname <> tg.lname OR src.age <> tg.age OR src.store_id <> tg.store_id"
+            f")"
+        )
+    )
+
+    staging = updates.select(
+        lit(0).alias("__key__"),
+        col("src.*"),
+        lit(start_date).cast("date").alias("start_date"),
+    ).unionByName(
+        source.alias("src").select(
+            lit(1).alias("__key__"),
+            col("src.*"),
+            lit(start_date).cast("date").alias("start_date"),
+        )
+    )
+
+    merge_expr = df.alias("tg").merge(
+        staging.alias("st"),
+        condition=(col("st.employee_id") == col("tg.employee_id")) & (col("st.__key__") == lit(1)),
+        clauses=[
+            WhenMatched(
+                condition=(col("tg.end_date") == lit(end_date))
+                & (
+                    (col("st.fname") != col("tg.fname"))
+                    | (col("st.lname") != col("tg.lname"))
+                    | (col("st.age") != col("tg.age"))
+                    | (col("st.store_id") != col("tg.store_id"))
+                )
+            ).update(
+                set_={
+                    "end_date": col("st.start_date"),
+                }
+            ),
```

Comment:
This is really cool. Like how easy it is to read (if you are familiar with PySpark).